File: android/app/src/main/java/com/vocahq/vocaphone/core/SnippetExpander.kt
package com.vocahq.vocaphone.core

import java.util.regex.Pattern

/**
 * Expands snippet triggers into their literal expansion text.
 *
 * Runs after [DictatedTranscript.finished], never before: the writing style
 * capitalizes sentence starts and would otherwise rewrite a snippet's literal
 * expansion (an email trigger must not come out "Me@example.com"). Matching
 * itself is case-insensitive, so capitalizing the source text first does not
 * break it.
 *
 * One combined regex over every trigger rather than a snippet-by-snippet
 * loop, so an expansion that happens to contain another trigger is never
 * re-expanded. Matches are applied in reverse order so an earlier match's
 * range stays valid while a later one is replaced.
 */
object SnippetExpander {

    fun expand(text: String, snippets: List<Snippet>): String {
        val active = snippets
            .filter { it.trigger.isNotBlank() }
            // Longer, more specific triggers win over a shorter one that could
            // be a substring of it ("my email" before "email").
            .sortedByDescending { it.trigger.length }
        if (active.isEmpty()) return text

        val pattern = Pattern.compile(
            active.joinToString("|") { "(${boundaryPattern(it.trigger)})" },
            Pattern.CASE_INSENSITIVE or Pattern.UNICODE_CASE or Pattern.UNICODE_CHARACTER_CLASS,
        )
        val matcher = pattern.matcher(text)
        val matches = mutableListOf<Triple<Int, Int, Int>>()
        while (matcher.find()) {
            val groupIndex = (1..active.size).first { matcher.group(it) != null } - 1
            matches += Triple(matcher.start(), matcher.end(), groupIndex)
        }
        if (matches.isEmpty()) return text

        val result = StringBuilder(text)
        for ((start, end, index) in matches.asReversed()) {
            result.replace(start, end, active[index].expansion)
        }
        return result.toString()
    }

    /**
     * `\b` on a side whose edge character is a word character; otherwise a
     * non-whitespace lookaround, so a punctuation-only trigger like "->" still
     * matches at the start or end of the string, where `\b` would not fire.
     */
    private fun boundaryPattern(trigger: String): String {
        val prefix = if (trigger.first().isWordChar()) "\\b" else "(?<!\\S)"
        val suffix = if (trigger.last().isWordChar()) "\\b" else "(?!\\S)"
        return prefix + Pattern.quote(trigger) + suffix
    }

    private fun Char.isWordChar(): Boolean = isLetterOrDigit() || this == '_'
}

File: ios/VocaPhoneShared/SnippetExpander.swift
import Foundation

protocol SnippetExpanding {
    func expand(in text: String, using snippets: [Snippet]) -> String
}

/// Expands every snippet trigger found in a finished transcript, in one pass.
///
/// One combined regex rather than looping snippet-by-snippet, so an expansion
/// that happens to contain another trigger is never itself re-expanded — a
/// second pass would see it as ordinary dictated text.
final class SnippetExpander: SnippetExpanding {
    func expand(in text: String, using snippets: [Snippet]) -> String {
        // A blank expansion is skipped rather than applied: the settings screen
        // will not save one, and a stored one would quietly delete its trigger
        // from every transcript, which is never what an empty field meant.
        let candidates = snippets.compactMap { snippet -> Snippet? in
            let trigger = snippet.trigger.trimmingCharacters(in: .whitespacesAndNewlines)
            let expansion = snippet.expansion.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !trigger.isEmpty, !expansion.isEmpty else { return nil }
            var trimmed = snippet
            trimmed.trigger = trigger
            return trimmed
        }
        guard !candidates.isEmpty else { return text }

        // Longest first, so "good morning team" wins over "good morning"
        // rather than the shorter trigger firing on a substring of it.
        let ordered = candidates.sorted { $0.trigger.count > $1.trigger.count }

        let pattern = ordered.map { "(\(boundaryPattern(for: $0.trigger)))" }.joined(separator: "|")
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive])
        else { return text }

        let nsText = text as NSString
        let matches = regex.matches(in: text, range: NSRange(location: 0, length: nsText.length))
        guard !matches.isEmpty else { return text }

        var result = text
        // Reverse order, so replacing a later match never invalidates the
        // range of a match still waiting to be replaced.
        for match in matches.reversed() {
            guard let groupIndex = matchedGroup(match), groupIndex - 1 < ordered.count,
                  let range = Range(match.range, in: result)
            else { continue }
            result.replaceSubrange(range, with: ordered[groupIndex - 1].expansion)
        }
        return result
    }

    private func matchedGroup(_ match: NSTextCheckingResult) -> Int? {
        for index in 1..<match.numberOfRanges where match.range(at: index).location != NSNotFound {
            return index
        }
        return nil
    }

    /// A trigger that starts or ends on a letter, digit, or underscore is
    /// bounded by ``word`` spelled out rather than by `\b`, so this matches
    /// the Android expander character for character — `\b` resolves against
    /// each engine's own idea of a word character, and the two do not agree.
    /// A trigger that starts or ends on punctuation — "->", ")))" — has no
    /// word character to bound, and `\b` would never fire beside it anyway;
    /// a non-whitespace lookaround matches it at either edge of the string as
    /// well as between words.
    private func boundaryPattern(for trigger: String) -> String {
        let escaped = NSRegularExpression.escapedPattern(for: trigger)
        let left = isWordCharacter(trigger.first) ? "(?<!\(Self.word))" : "(?<!\\S)"
        let right = isWordCharacter(trigger.last) ? "(?!\(Self.word))" : "(?!\\S)"
        return left + escaped + right
    }

    /// Letters and digits in any script, plus the underscore.
    private static let word = "[\\p{L}\\p{N}_]"

    private func isWordCharacter(_ character: Character?) -> Bool {
        guard let character else { return false }
        return character.isLetter || character.isNumber || character == "_"
    }
}
